kratos/tensor_adaptors/tensor_adaptor.h

```diff
@@ -0,0 +1,142 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Suneth Warnakulasuriya
+//
+
+#pragma once
+
+// System includes
+#include <string>
+#include <atomic>
+
+// External includes
+#include <span/span.hpp>
+
+// Project includes
+#include "includes/define.h"
+#include "includes/ublas_interface.h"
+#include "intrusive_ptr/intrusive_ptr.hpp"
+
+namespace Kratos {
+
+///@name Kratos Classes
+///@{
+
+/**
+ * @brief Base class or all the tensor adaptor types.
+ */
+template<class TContainerType, class TPrimitiveDataType>
+class KRATOS_API(KRATOS_CORE) TensorAdaptor {
+public:
+
+    ///@name Type definitions
+    ///@{
+
+    using Pointer = Kratos::intrusive_ptr<TensorAdaptor>;
+
+    using ConstPointer = Kratos::intrusive_ptr<const TensorAdaptor>;
+
+    using PrimitiveDataType = TPrimitiveDataType;
+
+    using ContainerType = TContainerType;
+
+    ///@}
+    ///@name Life cycle
+    ///@{
+
+    TensorAdaptor(typename TContainerType::Pointer pContainer) : mpContainer(pContainer) {};
+
+    virtual ~TensorAdaptor() = default;
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    /**
+     * @brief Fill the internal data from Kratos data structures
+     */
+    virtual void CollectData() = 0;
+
+    /**
+     * @brief Store internal data to the given TContainerType container.
+     *
+     */
+    virtual void StoreData() = 0;
+
+    typename TContainerType::Pointer GetContainer() const { return mpContainer; }
+
+    DenseVector<TPrimitiveDataType> MoveData() { return std::move(mData); }
+
+    Kratos::span<const TPrimitiveDataType> ViewData() const { return Kratos::span<const TPrimitiveDataType>(mData.data().begin(), mData.data().end()); }
```

Comment:
virtual

Reply:
the entire goal of making them virtual would be to use rutime polymorphism and to avoid variants. IT should come at negligible cost if we achieve it, and the code would be much cleaner